druid-shell/src/platform/x11/application.rs

```diff
@@ -539,18 +574,15 @@ fn poll_with_timeout(conn: &Rc<XCBConnection>, idle: RawFd, timeout: Instant) ->
                     // There is an X11 event ready to be handled.
                     break;
                 }
+                let now = Instant::now();
+                if now >= idle_timeout.min(timer_timeout.unwrap_or(idle_timeout)) {
+                    break;
+                }
```

Comment:
I think we want to break if `(past the timeout) || (past the idle timeout && there is an idle event)`. Maybe something like:
```suggestion
let now = Instant::now();
if timer_timeout.is_some() && now >= timer_timeout.unwrap() {
    break;
}
```

Reply:
> if (past the timeout) || (past the idle timeout && there is an idle event)

Let me rewrite that a bit:
* `now >= timer_timeout || now >= idle_timeout`
Next, if `now` is larger than one of them, then `now` is also larger than the smaller one of the two, so the above is equivalent to:
*  `now >= min(timer_timeout, idle_timeout)`
However, `idle_timeout` is an `Option`, so could be `None`. Let's handle this via `timer_timeout.unwrap_or(idle_timeout)`. There are two cases here:
* If there is a `timer_timeout`, this is equivalent to the above (take the smaller one of the two)
* If there is no `timer_timeout`, this does `now >= min(idle_timeout, idle_timeout)`, and obviously `min(a, a) == a`

I agree that this code is too clever, but I think it does what you want it to do. Hopefully, changing the `idle_timeout` logic to use a timer should simplify all of this, but... dunno, somehow I didn't want to do that.

----

For your suggested change: What happened to `idle_timeout`? It is no longer mentioned.